### DISCUSSION TITLE
Digest Authentication

### DISCUSSION THREAD

**Discussion by anilbolisetti** [lookfirst/sardine#429] [Q&A]
Does Sardine Library support Digest Authentication.
Have tried using SardineImpl.setCredentials setting AuthScope to Digest and i am getting the following exception:
com.github.sardine.impl.SardineException: status code: 401, reason phrase: Unexpected response (401 Unauthorized)

**Comment by lookfirst** [comment]
Hi there, this is a discussion, not an issue. Sardine supports auth.

https://github.com/lookfirst/sardine/wiki/UsageGuide#authentication

I suggest that you turn on debug/trace logging to better understand what is happening under the covers.

**Comment by anilbolisetti** [comment]
Hi,

I am trying to invoke SardineImpl.put passing inputStream.
Attached the code snippet:
SardineImpl sardineImpl = new SardineImpl();
			sardineImpl.enablePreemptiveAuthentication(url.getUrl());
			org.apache.http.client.CredentialsProvider provider = new BasicCredentialsProvider();
			provider.setCredentials(new AuthScope(AuthScope.ANY_HOST, -1, AuthScope.ANY_REALM, "Digest"), new UsernamePasswordCredentials(getUserName(), getPassword()));
			sardineImpl.setCredentials(provider);
			sardineImpl.put(url.getUrl(), is);
			sardineImpl.disablePreemptiveAuthentication();

It is successful with Basic authentication but fails with Digest authentication.
I am seeing the following exception:
FileServerDAO.createFolder() : com.github.sardine.impl.SardineException: status code: 401, reason phrase: Unexpected response (401 Unauthorized)
{FileServerDAO}{putFile}{APP1051}{WebDAV exception putting file on remote server. Connection reset}{External Message:Connection reset}

I have added the following appenders
<Property name="log-pattern">[%d{yyyy-MM-dd HH:mm:ss}][%t][%5p] %replace{%msg}{[\r\n]+}{(Encoded)}%n</Property>
<RollingFile name="root-log" fileName="${log-path}/webdav.log"
	                    filePattern="${log-path}/$${date:yyyy-MM}/AG-%d{MM-dd-yyyy}-%i.log.gz">
		        <PatternLayout>
		            <Pattern>${log-pattern}</Pattern>
		        </PatternLayout>
		        <Policies>
		            <TimeBasedTriggeringPolicy interval="1" modulate="true"/>
		            <SizeBasedTriggeringPolicy size="5000KB"/>
		            <DefaultRolloverStrategy max="10"/>
		        </Policies>
	        </RollingFile>
<Logger name="com.github" level="info" additivity="false">
	    <AppenderRef ref="root-log"/>
	</Logger>
	<logger name="org.apache.http.headers" level="info" additivity="false">
        <AppenderRef ref="root-log"/>
    </logger>
    <logger name="org.apache.http.wire" level="info" additivity="false">
        <AppenderRef ref="root-log"/>
    </logger>
No luck with getting extra logs from Sardine Classes.

Any help on sample code snippet which uses Digest authentication is appreciated.

**Reply by lookfirst** [reply]
I am guessing that `enablePreemptiveAuthentication` will require a host, not a url.

I also found this example in the HttpClient documentation:

https://github.com/apache/httpcomponents-client/blob/4.5.x/httpclient/src/examples/org/apache/http/examples/client/ClientPreemptiveDigestAuthentication.java

> No luck with getting extra logs from Sardine Classes.

https://github.com/lookfirst/sardine/blob/master/log4j.xml

**Reply by anilbolisetti** [reply]
I am guessing that enablePreemptiveAuthentication will require a host, not a url.

I also found this example in the HttpClient documentation:

https://github.com/apache/httpcomponents-client/blob/4.5.x/httpclient/src/examples/org/apache/http/examples/client/ClientPreemptiveDigestAuthentication.java

> This one is httpclient implementation. Does Sardine library provide any feature like this?

No luck with getting extra logs from Sardine Classes.

https://github.com/lookfirst/sardine/blob/master/log4j.xml

> I have added the entries from the above example to my jboss/log4j.xml as appenders and do not see any extra logs on exceptions.

**Reply by lookfirst** [reply]
Sardine exposes the HttpClient so you can override/implement whatever you want.

Sardine doesn't log much at all. You probably want HttpClient logs, not Sardine logs.

**Comment by anilbolisetti** [comment]
Property name="log-pattern">[%d{yyyy-MM-dd HH:mm:ss}][%t][%5p] %replace{%msg}{[\r\n]+}{(Encoded)}%n</Property
RollingFile name="root-log" fileName="${log-path}/webdav.log"
	                    filePattern="${log-path}/$${date:yyyy-MM}/AG-%d{MM-dd-yyyy}-%i.log.gz"
		        <PatternLayout
		            <Pattern>${log-pattern}</Pattern
		        </PatternLayout
		        <Policies
		            <TimeBasedTriggeringPolicy interval="1" modulate="true"/
		            <SizeBasedTriggeringPolicy size="5000KB"/
		            <DefaultRolloverStrategy max="10"/
		        </Policies
	        </RollingFile
<Logger name="com.github" level="info" additivity="false"
	    <AppenderRef ref="root-log"/
	</Logger
	<logger name="org.apache.http.headers" level="info" additivity="false"
        <AppenderRef ref="root-log"/
    </logger
    <logger name="org.apache.http.wire" level="info" additivity="false"
        <AppenderRef ref="root-log"/
    </logger

**Reply by lookfirst** [reply]
`logger name="org.apache.http.wire" level="info"`

probably want debug or trace there.
